scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+        self._maxtime = 3600
+        self._maxfuncall = 1e7
+        self._maxsteps = 500
+        self._temprestart = 0.1
+        self._qv = 2.62
+        self._usey = 1
+        self._xgas = 0.0
+        self._ygas = 0.0
+        self._ranbyx = 0.0
+        self._ranbyy = 0.0
+        self._sgas = -1.0
+        self._qa = -5.0  # Acceptance parameter
+        self.nblocal = 0
+        self._step_record = 0
+        self._nbfuncall = 0
+        self._extensive = False
+
+    def _judge_constraint(self):
+        # Not implemented yet
+        return True
+
+    def initialize(self):
+        """
+        Random coordinate generation in case given initial coordinates
+        are giving invalid objective value
+        """
+        self._nbfuncall = 0
+        self._indTrace = 0
+        if self._markovlength % self._dim != 0:
+            raise GenSARunnerException('Incorrect markov length.')
+        inconstraint = True
+        initerror = True
+        reinit_counter = 0
+        while(initerror):
+            if inconstraint:
+                if self._hasconstraint:
+                    inconstraint = self._judge_constraint()
+                    while(not inconstraint):
+                        self._coordin()
+                        inconstraint = self._judge_constraint()
+            self._energy(self._x)
+            if self._etot >= BIG_VALUE:
+                if reinit_counter >= MAX_REINIT_COUNT:
+                    initerror = False
+                    self._message = [(
+                        'Stopping algorithm because function '
+                        'create NaN or (+/-) inifinity values even with '
+                        'trying new random parameters')]
+                    raise GenSARunnerException(self._message)
+                for i in range(self._dim):
+                    self._x[i] = self._lower[i] + np.random.ranf() * (
+                        self._upper[i] - self._lower[i])
+                reinit_counter += 1
+            else:
+                initerror = False
+
+    def start_search(self):
+        """ Start annealing search process """
+        self.nblocal += 1
+        inconstraint = True
+        itnew = 0
+        emini_unchanged = True
+        emini_markov = 0.0
+        index_no_emini_update = 0
+        index_tol_emini_update = 1000
+        xminimarkov = np.zeros(self._dim)
+        etot_ls_ex = BIG_VALUE
+        # Here check that the function is smooth
+        self._starttime = time.time()
+        self._emini = np.array(self._etot)
+        self._xmini = np.array(self._x)
+        self._etot0 = np.array(self._etot)
+        if self._etot < self._emini:
+            self._emini = np.array(self._etot)
+            self._xmini = np.array(self._x)
+        self._etot0 = self._etot
+        if self._check_stopping():
+            self._stop_search()
+        self._step_record = 0
+        self._temp = self._tempsta
+        tolowtemp = True
+        while(tolowtemp):
```

Comment:
This is a big multiply nested loop. Is there any chance of factorising / simplifying it into sub functions, or something similar?